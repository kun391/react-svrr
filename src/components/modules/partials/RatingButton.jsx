import React from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import axios from 'axios';
import { API_URL } from 'constants/config';

class RatingButton extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      rating: this.props.rating || 0
    };

    this.options = [
      {value: 1, label: '1'},
      {value: 2, label: '2'},
      {value: 3, label: '3'},
      {value: 4, label: '4'},
      {value: 5, label: '5'},
      {value: 6, label: '6'},
      {value: 7, label: '7'},
      {value: 8, label: '8'},
      {value: 9, label: '9'},
      {value: 10, label: '10'}
    ];
  }

  ratingBook(data) {
    this.setState({
      rating: data.value
    });

    let params = {
      rating: data.value
    };

    axios({
      url: `${API_URL}/books/${this.props.id}/votes/`,
      method: 'PUT',
      data: params
    }).then(response => {
      alert('Rating successful.');

      if (typeof this.props.onChange === 'function') {
        this.props.onChange();
      }
    });
  }

  render() {
    return (
      <div className="rating-button">
        <div className="select-content rating">
          <span className="s-label">My rating:</span>
          <div className="select">
            <Select
              name="select-picker"
              value={this.state.rating}
              options={this.options}
              onChange={this.ratingBook.bind(this)}
              clearable={false}
              searchable={false}
              placeholder="---"
            />
          </div>
          <div className="clearfix"></div>
        </div>
      </div>
    );
  }
}

RatingButton.propTypes = {
  id: PropTypes.node,
  rating: PropTypes.number,
  onChange: PropTypes.func
};

export default RatingButton;
