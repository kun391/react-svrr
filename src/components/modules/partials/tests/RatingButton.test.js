/* global jest, describe, test, expect */
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import axios from 'axios'

import RatingButton from '../RatingButton'

jest.mock('axios', () => jest.fn())

axios.mockImplementation(() => Promise.resolve())

const getComponent = (props = {}) => shallow(<RatingButton {...props} />)
describe('RatingButton Component', () => {
  test('RatingButton should render as expected', () => {
    expect(toJson(getComponent())).toMatchSnapshot()
  })
  test('Should start with rating of 0 if no rating passed', () => {
    const component = getComponent({ id: 1, onChange: jest.fn() })
    expect(component.state().rating).toEqual(0)
  })
  test('ratingBook should appropriately rate the book', async () => {
    window.alert = jest.fn()
    const onChangeMock = jest.fn()
    const component = getComponent({ id: 1, onChange: onChangeMock })

    const data = {value: 5}
    await component.instance().ratingBook(data)

    expect(onChangeMock).toHaveBeenCalled()
  })
})
