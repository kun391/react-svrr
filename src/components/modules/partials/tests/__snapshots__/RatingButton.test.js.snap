// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingButton Component RatingButton should render as expected 1`] = `
<div
  className="rating-button"
>
  <div
    className="select-content rating"
  >
    <span
      className="s-label"
    >
      My rating:
    </span>
    <div
      className="select"
    >
      <Select
        arrowRenderer={[Function]}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        clearAllText="Clear all"
        clearRenderer={[Function]}
        clearValueText="Clear value"
        clearable={false}
        closeOnSelect={true}
        deleteRemoves={true}
        delimiter=","
        disabled={false}
        escapeClearsValue={true}
        filterOptions={[Function]}
        ignoreAccents={true}
        ignoreCase={true}
        inputProps={Object {}}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function]}
        multi={false}
        name="select-picker"
        noResultsText="No results found"
        onBlurResetsInput={true}
        onChange={[Function]}
        onCloseResetsInput={true}
        onSelectResetsInput={true}
        openOnClick={true}
        optionComponent={[Function]}
        options={
          Array [
            Object {
              "label": "1",
              "value": 1,
            },
            Object {
              "label": "2",
              "value": 2,
            },
            Object {
              "label": "3",
              "value": 3,
            },
            Object {
              "label": "4",
              "value": 4,
            },
            Object {
              "label": "5",
              "value": 5,
            },
            Object {
              "label": "6",
              "value": 6,
            },
            Object {
              "label": "7",
              "value": 7,
            },
            Object {
              "label": "8",
              "value": 8,
            },
            Object {
              "label": "9",
              "value": 9,
            },
            Object {
              "label": "10",
              "value": 10,
            },
          ]
        }
        pageSize={5}
        placeholder="---"
        removeSelected={true}
        required={false}
        rtl={false}
        scrollMenuIntoView={true}
        searchable={false}
        simpleValue={false}
        tabSelectsValue={true}
        trimFilter={true}
        value={0}
        valueComponent={[Function]}
        valueKey="value"
      />
    </div>
    <div
      className="clearfix"
    />
  </div>
</div>
`;
